In [2]:
import tensorflow as tf
import numpy as np

In [3]:
a = np.array(np.arange(1,31), dtype=np.float32).reshape([1,10,3])
print(a)

[[[ 1.  2.  3.]
  [ 4.  5.  6.]
  [ 7.  8.  9.]
  [10. 11. 12.]
  [13. 14. 15.]
  [16. 17. 18.]
  [19. 20. 21.]
  [22. 23. 24.]
  [25. 26. 27.]
  [28. 29. 30.]]]


上面的代码构建了1个1x10通道为3的直线数据，3个通道的数值分别为

| - | - | - | - | - | - | - | - | - | - |
|---|---|---|---|---|---|---|---|---|---|
| 1 | 4 | 7 | 10| 13| 16| 19| 22| 25| 28|

| - | - | - | - | - | - | - | - | - | - |
|---|---|---|---|---|---|---|---|---|---|
| 2 | 5 | 8 | 11| 14| 17| 20| 23| 26| 29|

| - | - | - | - | - | - | - | - | - | - |
|---|---|---|---|---|---|---|---|---|---|
| 3 | 6 | 9 | 12| 15| 18| 21| 24| 27| 30|

In [5]:
kernel = np.array(np.arange(1,46), dtype=np.float32).reshape([3,3,5])
print(kernel)

[[[ 1.  2.  3.  4.  5.]
  [ 6.  7.  8.  9. 10.]
  [11. 12. 13. 14. 15.]]

 [[16. 17. 18. 19. 20.]
  [21. 22. 23. 24. 25.]
  [26. 27. 28. 29. 30.]]

 [[31. 32. 33. 34. 35.]
  [36. 37. 38. 39. 40.]
  [41. 42. 43. 44. 45.]]]


上面的代码构建了5x3个1x3的矩阵，如下：

#### 第一组（对应input的3个channels）

|-|-|-|
|-|-|-|
|1|16|31|

|-|-|-|
|-|-|-|
|6|21|36|

|-|-|-|
|-|-|-|
|11|26|41|

#### 第二组（对应input的3个channels）

|-|-|-|
|-|-|-|
|2|17|32|

|-|-|-|
|-|-|-|
|7|22|37|

|-|-|-|
|-|-|-|
|12|27|42|

#### 第三组（对应input的3个channels）

|-|-|-|
|-|-|-|
|3|18|33|

|-|-|-|
|-|-|-|
|8|23|38|

|-|-|-|
|-|-|-|
|13|28|43|

#### 第四组（对应input的3个channels）

|-|-|-|
|-|-|-|
|4|19|34|

|-|-|-|
|-|-|-|
|9|24|39|

|-|-|-|
|-|-|-|
|14|29|44|

#### 第五组（对应input的3个channels）

|-|-|-|
|-|-|-|
|5|20|35|

|-|-|-|
|-|-|-|
|10|25|40|

|-|-|-|
|-|-|-|
|15|30|45|

**从上面可以看出，kernel卷积的大小（1x3）和input没有关系，是自己定义的。第2个参数input_channels=3为什么一定要和输入数据的channels一致呢？因为在我们需要从这层每次提取一个元素来组建上面的一组卷积矩阵，output_channels=5表示要准备好多少组这样的卷积矩阵，每组卷积矩阵学习一个特征，5表示希望学习到5种特征**

In [6]:
# 第3个参数为stride(不是strides复数)，而且值取整数，表示每次迭代向右移动几格
conv1d = tf.nn.conv1d(a, kernel, stride=1, padding='VALID')

In [7]:
with tf.Session() as sess:
    tf.global_variables_initializer().run()
    result = sess.run(conv1d)
    print(result)

[[[1245. 1290. 1335. 1380. 1425.]
  [1812. 1884. 1956. 2028. 2100.]
  [2379. 2478. 2577. 2676. 2775.]
  [2946. 3072. 3198. 3324. 3450.]
  [3513. 3666. 3819. 3972. 4125.]
  [4080. 4260. 4440. 4620. 4800.]
  [4647. 4854. 5061. 5268. 5475.]
  [5214. 5448. 5682. 5916. 6150.]]]


**输出Channel-1(特征1)：**

|-|-|-|-|-|-|-|-|
|-|-|-|-|-|-|-|-|
|1245|1812|2379|2946|3513|4080|4647|5214|

说明：conv1d每个channel之间的卷积结果是相加的，则整个计算过程为：点乘、加、加

**输出Channel-2(特征2)：**

|-|-|-|-|-|-|-|-|
|-|-|-|-|-|-|-|-|
|1290|1884|2478|3072|3666|4260|4854|5448|

**其它输出层(特征提取)的计算过程类似，自己推导就好。**
